### PR TITLE
Vesting improvements

### PR DESCRIPTION
**Pull request by 0xekez**
This significantly reworks the vesting contract to:

1. Fix an out-of-gas issue wherein a receiver could delegate to many validators and prevent the owner from being able to cancel it.
2. Add support for slashing.
3. Track bonded and unbonding tokens which fixes numerous issues with the old one where it would fail to distribute tokens and not be able to say much about why.

### CODE REVIEW

**Comment by 0xekez** [comment]
# slashing

Slashing can happen while tokens are staked to a validator, or while tokens are unbonding from a validator (ref: unbonding durations protect against long range attacks). Let's investigate how slashing impacts this contract.

In this contract we use two math formulas, $liquid(t)$ tells us the contract's current liquid token balance, and $claimable(t)$ tells us how many tokens may be claimed by the vest receiver. As the vest receiver selects which validators to delegate to, we expect that they will be penalized for slashing, and not the vest owner. This makes our formulas:

$$
liquid(t) := total - claimed(t) - staked(t) - slashed(t)
$$

$$
claimable(t) := vested(t) - claimed(t) - slashed(t)
$$

The Cosmos SDK does not provide a way for contracts to be notified when a slash happens, so let's consider what happens if a slash occurs and the contract does not know about it.

## slashed while staked

If tokens are slashed while they are staked and the contract does not know of it, it will make $staked(t) = staked(t) + slashed(t)$, as the contract will not know of the slash and thus will not deduct it from the staked balance. This means that $liquid$ will continue to return correct values.

$claimable$ on the other hand will report values $slashed(t)$ too large. this has different impacts depending on if the contract is canceled or not.

### slash while staked, contract open

The amount of funds to distribute is $min(liquid(t), claimable(t))$. This means that in the time following the slash claimable will be $slashed(t)$ too large and the receiver will be able to withdraw "too much"; however, once the vest completes $liquid(t)$ being correct will stop any claiming of funds $\gt total$.

### slashed while staked, contract cancelled

When a contract is canceled currently liquid funds are sent to the vest receiver up to the amount that they may claim and, $total$ is set to $vested(t)$.

$$
settle = min(claimable(t), liquid(t))
$$

Because $claimable(t)$ is $slashed(t)$ too large, **a closed contract with a slash may distribute too many tokens to the vest receiver**. This makes the owner bear the cost of shashing.

## slashed while unbonding

If tokens are slashed while they are unbonding this will make $liquid(t)$ $slashed(t)$ too large, and $claimable(t)$ $slashed(t)$ too large.

### slashed while unbonding, contract open

This has the same outcome as being slashed while bonded and open as the factory pattern will prevent a vesting contract from being able to distribute more tokens than it has (x/bank to the rescue).

### slashed while unbonding, contract closed

The contract will not be closable as the overestimate of $liquid(t)$ will cause the contract to attempt to settle more funds than it has (x/bank will error).

**Comment by JakeHartnell** [comment]
Looking GREAT. Made a PR with some minor clean up: #654 

IMO this G2G,  though I don't understand why the owner recieves staking rewards... or to rephrase, I think that the vestee should get rewards up until the contract being canceled. These could be quite substantial if the vesting contract is large. Perhaps we can find a way to address this?

**Comment by JakeHartnell** [comment]
Wish list. Wormhole is galaxy brain amazing btw. Have many other places I want to use it?

Wormhole:
- Ability to zero a value at time t
- What are the state implications of this? Is there a way to clear history?

For stake tracker:
- Could make it it's own package? It's great.
- Would be maybe nice to add rewards? This way we could easily use it for mesh though I would use it right now. Related question: does undelegating or redelegating auto withdraw staking rewards? We could thus track them in the stake tracker?

**Review by bekauz — APPROVED** [review]

File: contracts/external/cw-vesting/src/contract.rs [hunk 1]
```diff
@@ -1,84 +1,87 @@
...
-        UncheckedDenom::Native(ref denom) => {
-            // Validate all vesting payment params
-            let checked_params = msg.params.clone().into_checked(deps.as_ref())?;
-
-            // Check amount sent matches amount in vesting payment
-            if checked_params.amount != must_pay(&info, denom)? {
-                return Err(ContractError::AmountDoesNotMatch);
+    let denom = msg.denom.into_checked(deps.as_ref())?;
+    let recipient = deps.api.addr_validate(&msg.recipient)?;
+    let start_time = msg.start_time.unwrap_or(env.block.time);
```
[the remark this answers is not in the document]
Comment (0xekez) [reply contracts/external/cw-vesting/src/contract.rs:34]: my thought was that a start time in the past is actually alright. for example, say you wanted to start a payment for an employee, but the DAO proposal ended up taking forever to pass and it wasn't until the end of the month that you got it executed. in that case the contract will work fine. i've added a test for this too. thanks for pointing it out!

[the remark this answers is not in the document]
Comment (0xekez) [reply contracts/external/cw-vesting/src/contract.rs:34]: my thought was that a start time in the past is actually alright. for example, say you wanted to start a payment for an employee, but the DAO proposal ended up taking forever to pass and it wasn't until the end of the month that you got it executed. in that case the contract will work fine. i've added a test for this too. thanks for pointing it out!

File: contracts/external/cw-vesting/src/vesting.rs [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+    pub title: String,
+    pub description: Option<String>,
+}
+
+#[cw_serde]
+pub enum Status {
+    Unfunded,
+    Funded,
+    Canceled {
+        /// owner_withdrawable(t). This is monotonicly decreasing and
```
[the remark this answers is not in the document]
Comment (0xekez) [reply contracts/external/cw-vesting/src/vesting.rs:35]: thank you :)

File: packages/cw-wormhole/README.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+```rust
+use cosmwasm_std::{testing::mock_dependencies, Uint128};
+use cw_wormhole::Wormhole;
+let storage = &mut mock_dependencies().storage;
+let w: Wormhole<Addr, Uint128> = Wormhole::new("ns");
+let key = Addr::unchecked("violet");
+
+// increment the value by one at time 10.
+fm.increment(storage, key.clone(), 10, Uint128::new(1))
```
[the remark this answers is not in the document]
Comment (0xekez) [reply packages/cw-wormhole/README.md:14]: haha you are the best. good catch 🙏

[the remark this answers is not in the document]
Comment (0xekez) [reply packages/cw-wormhole/README.md:14]: added the README code to be tested as part of our unit tests so this won't happen again.

File: contracts/external/cw-vesting/src/vesting.rs [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+                        .denom
+                        .get_transfer_to_message(&vesting.recipient, to_vestee)?,
+                );
+            }
+
+            Ok(msgs)
+        }
+    }
+
+    pub fn withdraw_canceled(
```
[the remark this answers is not in the document]
Comment (0xekez) [reply contracts/external/cw-vesting/src/vesting.rs:258]: great feedback. thank you. i've renamed it to `withdraw_canceled_payment`.

**Review by JakeHartnell — APPROVED** [review]
LFG!

**Pull request merged**